File: graphlib-2/go/lib/traversal.go
package lib

import (
	"sort"
)

// BFSResult contains the results of a BFS traversal
type BFSResult struct {
	Order  []string          `json:"order"`
	Levels map[string]int    `json:"levels"`
	Parent map[string]string `json:"parent"`
}

// BFS performs breadth-first search from the start node
func (g *Graph) BFS(start string) *BFSResult {
	g.mu.RLock()
	defer g.mu.RUnlock()

	if !g.Nodes[start] {
		return &BFSResult{
			Order:  []string{},
			Levels: make(map[string]int),
			Parent: make(map[string]string),
		}
	}

	result := &BFSResult{
		Order:  make([]string, 0),
		Levels: make(map[string]int),
		Parent: make(map[string]string),
	}

	visited := make(map[string]bool)
	queue := []string{start}
	visited[start] = true
	result.Levels[start] = 0

	for len(queue) > 0 {
		node := queue[0]
		queue = queue[1:]
		result.Order = append(result.Order, node)

		// Get neighbors in sorted order for deterministic traversal
		neighbors := make([]string, 0)
		for neighbor := range g.Adj[node] {
			neighbors = append(neighbors, neighbor)
		}
		sort.Strings(neighbors)

		for _, neighbor := range neighbors {
			if !visited[neighbor] {
				visited[neighbor] = true
				queue = append(queue, neighbor)
				result.Levels[neighbor] = result.Levels[node] + 1
				result.Parent[neighbor] = node
			}
		}
	}

	return result
}

// DFSResult contains the results of a DFS traversal
type DFSResult struct {
	Order     []string          `json:"order"`
	Discovery map[string]int    `json:"discovery"`
	Finish    map[string]int    `json:"finish"`
	Parent    map[string]string `json:"parent"`
}

// DFS performs depth-first search from the start node
func (g *Graph) DFS(start string) *DFSResult {
	g.mu.RLock()
	defer g.mu.RUnlock()

	if !g.Nodes[start] {
		return &DFSResult{
			Order:     []string{},
			Discovery: make(map[string]int),
			Finish:    make(map[string]int),
			Parent:    make(map[string]string),
		}
	}

	result := &DFSResult{
		Order:     make([]string, 0),
		Discovery: make(map[string]int),
		Finish:    make(map[string]int),
		Parent:    make(map[string]string),
	}

	visited := make(map[string]bool)
	time := 0

	var dfsVisit func(node string)
	dfsVisit = func(node string) {
		visited[node] = true
		result.Discovery[node] = time
		time++
		result.Order = append(result.Order, node)

		// Get neighbors in sorted order for deterministic traversal
		neighbors := make([]string, 0)
		for neighbor := range g.Adj[node] {
			neighbors = append(neighbors, neighbor)
		}
		sort.Strings(neighbors)

		for _, neighbor := range neighbors {
			if !visited[neighbor] {
				result.Parent[neighbor] = node
				dfsVisit(neighbor)
			}
		}

		result.Finish[node] = time
		time++
	}

	dfsVisit(start)
	return result
}

File: graphlib-3/py/graphlib/traversal.py
"""BFS and DFS traversal implementations."""
from typing import Dict, List, Any
from collections import deque
from .types import get_graph


def bfs(graph_id: str, start_node: str) -> Dict[str, Any]:
    """Breadth-first search traversal."""
    graph = get_graph(graph_id)
    if graph is None:
        return {"order": [], "levels": {}, "parent": {}}

    if start_node not in graph.nodes:
        return {"order": [], "levels": {}, "parent": {}}

    order = []
    levels = {}
    parent = {}
    visited = set()

    queue = deque([(start_node, 0)])
    visited.add(start_node)

    while queue:
        node, level = queue.popleft()
        order.append(node)
        levels[node] = level

        neighbors = sorted(graph.adjacency.get(node, {}).keys())
        for neighbor in neighbors:
            if neighbor not in visited:
                visited.add(neighbor)
                parent[neighbor] = node
                queue.append((neighbor, level + 1))

    return {"order": order, "levels": levels, "parent": parent}


def dfs(graph_id: str, start_node: str) -> Dict[str, Any]:
    """Depth-first search traversal."""
    graph = get_graph(graph_id)
    if graph is None:
        return {"order": [], "discovery": {}, "finish": {}, "parent": {}}

    if start_node not in graph.nodes:
        return {"order": [], "discovery": {}, "finish": {}, "parent": {}}

    order = []
    discovery = {}
    finish = {}
    parent = {}
    visited = set()
    time = [0]

    def dfs_visit(node: str):
        visited.add(node)
        discovery[node] = time[0]
        time[0] += 1
        order.append(node)

        neighbors = sorted(graph.adjacency.get(node, {}).keys())
        for neighbor in neighbors:
            if neighbor not in visited:
                parent[neighbor] = node
                dfs_visit(neighbor)

        finish[node] = time[0]
        time[0] += 1

    dfs_visit(start_node)

    return {"order": order, "discovery": discovery, "finish": finish, "parent": parent}
